File: packages/host/src/recording.ts
import { decodeInputBatch } from "@dopejs/pingo-editing";
import { decodeMutationBatch } from "@dopejs/pingo-reconciler";

import { decodeSystemTextMetricBatch } from "./system-text-metrics";

import {
  ABI_VERSION,
  MAX_RECORDING_BYTES,
  MAX_RECORDING_RECORDS,
  PROTOCOL_ALIGNMENT,
  RECORD_HEADER_BYTES,
  RECORDING_MAGIC,
  RecordingRecordKind,
  STREAM_HEADER_BYTES,
  INSTRUCTION_FLAG_MASK,
  INSTRUCTION_FLAG_OPTIONAL,
  INSTRUCTION_LENGTH_ESCAPE,
  MINIMUM_READABLE_ABI_VERSION,
} from "./generated";

const MUTATION_RECORD_KIND = Number(RecordingRecordKind.Mutation);
const INPUT_RECORD_KIND = Number(RecordingRecordKind.Input);
const SYSTEM_TEXT_METRICS_RECORD_KIND = Number(RecordingRecordKind.SystemTextMetrics);
const ANIMATION_FRAME_RECORD_KIND = Number(RecordingRecordKind.AnimationFrame);

/** Explicit privacy classification required before binary data may be retained. */
export type ReplayDataClassification = "recordable" | "sensitive";

/** One exact nested stream in observed order. */
export type ReplayRecord =
  | { readonly type: "mutation"; readonly bytes: Uint8Array }
  | { readonly type: "input"; readonly bytes: Uint8Array }
  | { readonly type: "systemTextMetrics"; readonly bytes: Uint8Array }
  | { readonly type: "animationFrame"; readonly elapsedMicros: bigint };

/** A recursively validated deterministic replay recording. */
export interface ReplayRecording {
  readonly records: readonly ReplayRecord[];
}

/** Malformed, oversized, unsupported, or privacy-unsafe recording operation. */
export class ReplayRecordingError extends Error {
  public constructor(message: string, options?: ErrorOptions) {
    super(message, options);
    this.name = "ReplayRecordingError";
  }
}

/** Encodes a canonical recording after validating every nested transaction. */
export function encodeReplayRecording(recording: ReplayRecording): Uint8Array {
  if (recording.records.length > MAX_RECORDING_RECORDS) fail("record count exceeds limit");
  let length: number = STREAM_HEADER_BYTES;
  for (const record of recording.records) {
    validateRecord(record);
    length = checkedAdd(length, RECORD_HEADER_BYTES);
    length = checkedAdd(length, payloadLength(record));
    if (length > MAX_RECORDING_BYTES) fail("recording exceeds maximum size");
  }

  const output = new Uint8Array(length);
  const view = new DataView(output.buffer);
  view.setUint32(0, RECORDING_MAGIC, true);
  view.setUint16(4, ABI_VERSION, true);
  view.setUint16(6, STREAM_HEADER_BYTES, true);
  view.setUint32(8, length, true);
  view.setUint32(12, recording.records.length, true);
  let offset = STREAM_HEADER_BYTES;
  for (const record of recording.records) {
    output[offset] = kindFor(record);
    // Records carry their length like every other instruction, so a reader that
    // does not know the kind can step over one instead of refusing the file.
    const payloadBytes = payloadLength(record);
    const total = RECORD_HEADER_BYTES + payloadBytes;
    const words = total / PROTOCOL_ALIGNMENT;
    if (words >= INSTRUCTION_LENGTH_ESCAPE) fail("recording record is too large to frame");
    view.setUint16(offset + 2, words, true);
    view.setUint32(offset + 4, payloadBytes, true);
    offset += RECORD_HEADER_BYTES;
    if (record.type === "animationFrame") {
      view.setBigUint64(offset, record.elapsedMicros, true);
    } else {
      output.set(record.bytes, offset);
    }
    offset += payloadBytes;
  }
  return output;
}

/** Decodes all records and nested streams before returning any replayable data. */
export function decodeReplayRecording(input: Uint8Array): ReplayRecording {
  if (input.byteLength > MAX_RECORDING_BYTES) fail("recording exceeds maximum size");
  if (input.byteLength % PROTOCOL_ALIGNMENT !== 0) fail("recording is not aligned");
  const reader = new RecordingReader(input);
  if (reader.u32() !== RECORDING_MAGIC) fail("wrong recording magic");
  // Newer producers stay readable through the self-describing instruction
  // framing; anything older than it cannot be stepped through safely.
  if (reader.u16() < MINIMUM_READABLE_ABI_VERSION) fail("unsupported recording ABI version");
  if (reader.u16() !== STREAM_HEADER_BYTES) fail("invalid recording header length");
  if (reader.u32() !== input.byteLength) fail("declared recording length does not match input");
  const declaredCount = reader.u32();
  if (declaredCount > MAX_RECORDING_RECORDS) fail("record count exceeds limit");
  if (declaredCount > Math.floor(reader.remaining / RECORD_HEADER_BYTES)) {
    fail("record count cannot fit in remaining bytes");
  }

  const records: ReplayRecord[] = [];
  let seenRecords = 0;
  while (reader.remaining > 0) {
    const start = reader.offset;
    const kind = reader.u8();
    const flags = reader.u8();
    if ((flags & ~INSTRUCTION_FLAG_MASK) !== 0) fail("unsupported recording flags");
    // A record header now carries the instruction length like every other
    // stream, so a reader that does not know the record kind can step over it.
    const words = reader.u16();
    const total = words === INSTRUCTION_LENGTH_ESCAPE ? reader.u32() : words * PROTOCOL_ALIGNMENT;
    const length = reader.u32();
    if (length % PROTOCOL_ALIGNMENT !== 0) fail("nested stream is not aligned");
    const bytes = reader.bytes(length);
    seenRecords += 1;
    if (reader.offset !== start + total) fail("record length does not match its payload");
    if (!(kind in RecordingRecordKind)) {
      if ((flags & INSTRUCTION_FLAG_OPTIONAL) === 0) fail("unknown recording record kind");
      continue;
    }
    const record = recordFor(kind, bytes);
    validateRecord(record);
    records.push(record);
  }
  if (seenRecords !== declaredCount) fail("record count does not match input");
  return { records };
}

/** Exact-order callbacks used by browser-free and incident replay tools. */
export interface ReplayHandlers {
  readonly mutation: (bytes: Uint8Array) => void;
  readonly input: (bytes: Uint8Array) => void;
  readonly systemTextMetrics: (bytes: Uint8Array) => void;
  readonly animationFrame: (elapsedMicros: bigint) => void;
}

/** Validates then replays a complete recording in exact observed order. */
export function replayRecording(input: Uint8Array, handlers: ReplayHandlers): void {
  const recording = decodeReplayRecording(input);
  for (const record of recording.records) {
    if (record.type === "animationFrame") handlers.animationFrame(record.elapsedMicros);
    else handlers[record.type](record.bytes.slice());
  }
}

/** Bounded in-memory recorder that never retains explicitly sensitive transactions. */
export class BinaryReplayRecorder {
  readonly #records: ReplayRecord[] = [];
  #encodedBytes: number = STREAM_HEADER_BYTES;

  /** Number of recordable transactions retained in memory. */
  public get size(): number {
    return this.#records.length;
  }

  /** Captures one exact Mutation Stream, or safely skips sensitive content. */
  public captureMutation(bytes: Uint8Array, classification: ReplayDataClassification): boolean {
    return this.capture({ type: "mutation", bytes }, classification);
  }

  /** Captures one exact Input Stream, or safely skips password/private content. */
  public captureInput(bytes: Uint8Array, classification: ReplayDataClassification): boolean {
    return this.capture({ type: "input", bytes }, classification);
  }

  /** Captures one exact browser system-font metric cache delta. */
  public captureSystemTextMetrics(
    bytes: Uint8Array,
    classification: ReplayDataClassification,
  ): boolean {
    return this.capture({ type: "systemTextMetrics", bytes }, classification);
  }

  /** Captures one exact, non-sensitive Core logical-clock delta. */
  public captureAnimationFrame(elapsedMicros: bigint): boolean {
    return this.capture({ type: "animationFrame", elapsedMicros }, "recordable");
  }

  /** Exports a detached versioned archive without exposing retained mutable buffers. */
  public export(): Uint8Array {
    return encodeReplayRecording({ records: this.#records });
  }

  /** Removes all retained diagnostic data from this recorder. */
  public clear(): void {
    this.#records.length = 0;
    this.#encodedBytes = STREAM_HEADER_BYTES;
  }

  private capture(record: ReplayRecord, classification: ReplayDataClassification): boolean {
    if (classification === "sensitive") return false;
    if (classification !== "recordable") fail("unknown replay data classification");
    validateRecord(record);
    if (this.#records.length >= MAX_RECORDING_RECORDS) fail("record count exceeds limit");
    const nextBytes = checkedAdd(
      this.#encodedBytes,
      checkedAdd(RECORD_HEADER_BYTES, payloadLength(record)),
    );
    if (nextBytes > MAX_RECORDING_BYTES) fail("recording exceeds maximum size");
    this.#records.push(
      record.type === "animationFrame" ? record : { ...record, bytes: record.bytes.slice() },
    );
    this.#encodedBytes = nextBytes;
    return true;
  }
}

class RecordingReader {
  readonly #input: Uint8Array;
  readonly #view: DataView;
  #offset = 0;

  public constructor(input: Uint8Array) {
    this.#input = input;
    this.#view = new DataView(input.buffer, input.byteOffset, input.byteLength);
  }

  public get remaining(): number {
    return this.#input.byteLength - this.#offset;
  }

  public get offset(): number {
    return this.#offset;
  }

  public u8(): number {
    this.require(1);
    return this.#view.getUint8(this.#offset++);
  }

  public u16(): number {
    this.require(2);
    const value = this.#view.getUint16(this.#offset, true);
    this.#offset += 2;
    return value;
  }

  public u32(): number {
    this.require(4);
    const value = this.#view.getUint32(this.#offset, true);
    this.#offset += 4;
    return value;
  }

  public zeroes(length: number): void {
    if (this.bytes(length).some((byte) => byte !== 0))
      fail("reserved recording bytes must be zero");
  }

  public bytes(length: number): Uint8Array {
    this.require(length);
    const result = this.#input.slice(this.#offset, this.#offset + length);
    this.#offset += length;
    return result;
  }

  private require(length: number): void {
    if (!Number.isSafeInteger(length) || length < 0 || length > this.remaining) {
      fail("truncated recording");
    }
  }
}

function validateRecord(record: ReplayRecord): void {
  try {
    if (record.type === "mutation") decodeMutationBatch(record.bytes);
    else if (record.type === "input") decodeInputBatch(record.bytes);
    else if (record.type === "systemTextMetrics") decodeSystemTextMetricBatch(record.bytes);
    else if (record.elapsedMicros < 0n || record.elapsedMicros > 0xffff_ffff_ffff_ffffn) {
      fail("animation frame delta is outside u64 range");
    }
  } catch (cause) {
    throw new ReplayRecordingError(`invalid nested ${record.type} stream`, { cause });
  }
}

function kindFor(record: ReplayRecord): RecordingRecordKind {
  if (record.type === "mutation") return RecordingRecordKind.Mutation;
  if (record.type === "input") return RecordingRecordKind.Input;
  if (record.type === "systemTextMetrics") return RecordingRecordKind.SystemTextMetrics;
  return RecordingRecordKind.AnimationFrame;
}

function recordFor(kind: number, bytes: Uint8Array): ReplayRecord {
  if (kind === MUTATION_RECORD_KIND) return { type: "mutation", bytes };
  if (kind === INPUT_RECORD_KIND) return { type: "input", bytes };
  if (kind === SYSTEM_TEXT_METRICS_RECORD_KIND) return { type: "systemTextMetrics", bytes };
  if (kind === ANIMATION_FRAME_RECORD_KIND) {
    if (bytes.byteLength !== 8) fail("animation frame payload must be eight bytes");
    return {
      type: "animationFrame",
      elapsedMicros: new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength).getBigUint64(
        0,
        true,
      ),
    };
  }
  return fail(`unknown recording record kind ${String(kind)}`);
}

function payloadLength(record: ReplayRecord): number {
  return record.type === "animationFrame" ? 8 : record.bytes.byteLength;
}

function checkedAdd(left: number, right: number): number {
  const result = left + right;
  if (!Number.isSafeInteger(result)) fail("recording size overflow");
  return result;
}

function fail(message: string): never {
  throw new ReplayRecordingError(message);
}

File: core/doper-abi/src/recording.rs
use crate::codec::{
    Reader, Writer, read_header, read_instruction_header, validate_encode_instruction_count,
};
use crate::{
    AbiError, InputBatch, MAX_RECORDING_BYTES, MAX_RECORDING_RECORDS, MutationBatch,
    RECORD_HEADER_BYTES, RECORDING_MAGIC, RecordingRecordKind, StreamKind,
};

/// One exact binary transaction in an ordered deterministic replay recording.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum ReplayRecord {
    /// A complete Shell-to-Core Mutation Stream.
    Mutation(Vec<u8>),
    /// A complete Host-to-Core Input Stream.
    Input(Vec<u8>),
}

impl ReplayRecord {
    fn kind(&self) -> RecordingRecordKind {
        match self {
            Self::Mutation(_) => RecordingRecordKind::Mutation,
            Self::Input(_) => RecordingRecordKind::Input,
        }
    }

    fn bytes(&self) -> &[u8] {
        match self {
            Self::Mutation(bytes) | Self::Input(bytes) => bytes,
        }
    }

    fn validate(&self) -> Result<(), AbiError> {
        match self {
            Self::Mutation(bytes) => MutationBatch::decode(bytes).map(|_| ()),
            Self::Input(bytes) => InputBatch::decode(bytes).map(|_| ()),
        }
    }
}

/// Versioned archive preserving the exact order and bytes of mutation and input transactions.
#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct ReplayRecording {
    /// Validated records in observed order.
    pub records: Vec<ReplayRecord>,
}

impl ReplayRecording {
    /// Decodes and recursively validates an untrusted recording before returning any record.
    pub fn decode(bytes: &[u8]) -> Result<Self, AbiError> {
        let mut reader = Reader::new(bytes);
        let declared_count = read_header(&mut reader, RECORDING_MAGIC, MAX_RECORDING_BYTES)?;
        if declared_count > MAX_RECORDING_RECORDS {
            return Err(AbiError::InstructionCountTooLarge {
                declared: declared_count,
                maximum: MAX_RECORDING_RECORDS,
            });
        }
        let maximum_count = u32::try_from(reader.remaining() / RECORD_HEADER_BYTES)
            .map_err(|_| AbiError::ArithmeticOverflow)?;
        if declared_count > maximum_count {
            return Err(AbiError::InstructionCountTooLarge {
                declared: declared_count,
                maximum: maximum_count,
            });
        }

        let capacity = usize::try_from(declared_count).map_err(|_| AbiError::ArithmeticOverflow)?;
        let mut records = Vec::with_capacity(capacity);
        while reader.remaining() != 0 {
            let (offset, raw_kind, _) = read_instruction_header(&mut reader)?;
            let kind = RecordingRecordKind::from_u8(raw_kind).ok_or(AbiError::UnknownOpcode {
                stream: StreamKind::Recording,
                opcode: raw_kind,
                offset,
            })?;
            let length =
                usize::try_from(reader.read_u32()?).map_err(|_| AbiError::ArithmeticOverflow)?;
            if !length.is_multiple_of(crate::PROTOCOL_ALIGNMENT) {
                return Err(AbiError::Misaligned {
                    offset: reader.offset().saturating_add(length),
                });
            }
            let payload = reader.read_bytes(length)?.to_vec();
            let record = match kind {
                RecordingRecordKind::Mutation => {
                    MutationBatch::decode(&payload)?;
                    ReplayRecord::Mutation(payload)
                }
                RecordingRecordKind::Input => {
                    InputBatch::decode(&payload)?;
                    ReplayRecord::Input(payload)
                }
            };
            records.push(record);
        }
        let actual = u32::try_from(records.len()).map_err(|_| AbiError::ArithmeticOverflow)?;
        if actual != declared_count {
            return Err(AbiError::InstructionCountMismatch {
                declared: declared_count,
                actual,
            });
        }
        Ok(Self { records })
    }

    /// Encodes a canonical recording after recursively validating all nested streams.
    pub fn encode(&self) -> Result<Vec<u8>, AbiError> {
        validate_encode_instruction_count(self.records.len(), 0, MAX_RECORDING_RECORDS)?;
        let mut writer = Writer::new(RECORDING_MAGIC);
        for record in &self.records {
            record.validate()?;
            let payload = record.bytes();
            let payload_length =
                u32::try_from(payload.len()).map_err(|_| AbiError::ArithmeticOverflow)?;
            writer.instruction(record.kind() as u8, 0);
            writer.u32(payload_length);
            writer.bytes(payload);
        }
        writer.finish(MAX_RECORDING_BYTES)
    }
}

#[cfg(test)]
mod tests {
    use proptest::prelude::*;

    use super::*;
    use crate::{
        InputCommand, InputInstruction, Mutation, MutationInstruction, NULL_NODE_ID, NodeKind,
    };

    fn mutation(frame_seq: u32) -> Vec<u8> {
        MutationBatch {
            frame_seq,
            instructions: vec![MutationInstruction {
                flags: 0,
                mutation: Mutation::CreateNode {
                    node_id: 1 << 20,
                    kind: NodeKind::Root,
                    parent: NULL_NODE_ID,
                    before_sibling: NULL_NODE_ID,
                },
            }],
        }
        .encode()
        .expect("mutation")
    }

    fn input(frame_seq: u32) -> Vec<u8> {
        InputBatch {
            frame_seq,
            instructions: vec![InputInstruction {
                flags: 0,
                command: InputCommand::Insert {
                    node_id: 1 << 20,
                    base_revision: 7,
                    text: "你".to_owned(),
                },
            }],
        }
        .encode()
        .expect("input")
    }

    #[test]
    fn preserves_exact_nested_bytes_and_observed_order() {
        let recording = ReplayRecording {
            records: vec![
                ReplayRecord::Mutation(mutation(1)),
                ReplayRecord::Input(input(2)),
                ReplayRecord::Mutation(mutation(3)),
            ],
        };
        let bytes = recording.encode().expect("recording");
        assert_eq!(ReplayRecording::decode(&bytes), Ok(recording));
    }

    #[test]
    fn rejects_invalid_envelopes_and_nested_streams() {
        let mut bytes = ReplayRecording {
            records: vec![ReplayRecord::Mutation(mutation(1))],
        }
        .encode()
        .expect("recording");

        let mut unknown = bytes.clone();
        unknown[16] = 0xff;
        assert!(matches!(
            ReplayRecording::decode(&unknown),
            Err(AbiError::UnknownOpcode {
                stream: StreamKind::Recording,
                opcode: 0xff,
                offset: 16,
            })
        ));

        let mut hostile_count = bytes.clone();
        hostile_count[12..16].copy_from_slice(&u32::MAX.to_le_bytes());
        assert!(matches!(
            ReplayRecording::decode(&hostile_count),
            Err(AbiError::InstructionCountTooLarge { .. })
        ));

        bytes[24] = 0;
        assert!(ReplayRecording::decode(&bytes).is_err());
        assert!(
            ReplayRecording {
                records: vec![ReplayRecord::Input(vec![1, 2, 3, 4])],
            }
            .encode()
            .is_err()
        );
    }

    proptest! {
        #[test]
        fn arbitrary_bytes_never_panic(bytes in prop::collection::vec(any::<u8>(), 0..4096)) {
            let _ = ReplayRecording::decode(&bytes);
        }
    }
}
